Bugsnag/KSCrash/Source/KSCrash/Recording/Tools/BSG_KSLogger.c

```diff
@@ -147,88 +77,81 @@ bool bsg_kslog_setLogFilename(const char *filename, bool overwrite) {
-    if (g_file != NULL && bsg_g_file != stdout && bsg_g_file != stderr &&
-        bsg_g_file != stdin) {
-        fclose(g_file);
+__printflike(1, 2)
+void bsg_i_kslog_logCBasic(const char *const fmt, ...) {
+    unlikely_if(!fmt) {
+        return;
     }
-    bsg_g_file = file;
-}
-
-static void writeToLog(const char *const str) {
-    unlikely_if(g_file == NULL) { bsg_g_file = stdout; }
-
-    fprintf(g_file, "%s", str);
-}
 
-static inline void writeFmtArgsToLog(const char *fmt, va_list args) {
-    unlikely_if(g_file == NULL) { bsg_g_file = stdout; }
+    char buf[BSG_KSLOGGER_CBufferSize];
+    const int size = sizeof(buf);
 
-    if (fmt == NULL) {
-        writeToLog("(null)");
-    } else {
-        vfprintf(g_file, fmt, args);
+    va_list args;
+    va_start(args, fmt);
+    int len = vsnprintf(buf, size, fmt, args);
+    if (len > size) {
+        len = size;
     }
-}
-
-static inline void flushLog(void) { fflush(g_file); }
+    va_end(args);
 
-bool bsg_kslog_setLogFilename(const char *filename, bool overwrite) {
-    if (filename == NULL) {
-        setLogFD(stdout);
-        return true;
+    if (len < 0) {
+        return;
     }
 
-    FILE *file = fopen(filename, overwrite ? "wb" : "ab");
-    unlikely_if(file == NULL) {
-        writeFmtToLog("KSLogger: Could not open %s: %s", filename,
-                      strerror(errno));
-        return false;
+    if (len < size - 1) {
+        buf[len++] = '\n';
+    } else {
+        buf[size - 1] = '\n';
     }
 
-    setLogFD(file);
-    return true;
-}
-
-#endif
-
-// ===========================================================================
-#pragma mark - C -
-// ===========================================================================
-
-__printflike(1, 2)
-void bsg_i_kslog_logCBasic(const char *const fmt, ...) {
-    va_list args;
-    va_start(args, fmt);
-    writeFmtArgsToLog(fmt, args);
-    va_end(args);
-    writeToLog("\n");
-    flushLog();
+    write(bsg_g_fd, buf, (size_t)len);
 }
 
 __printflike(5, 6)
 void bsg_i_kslog_logC(const char *const level, const char *const file,
                       const int line, const char *const function,
                       const char *const fmt, ...) {
-    writeFmtToLog("%s: %s:%u: %s(): ", level, lastPathEntry(file), line,
-                  function);
-    va_list args;
-    va_start(args, fmt);
-    writeFmtArgsToLog(fmt, args);
-    va_end(args);
-    writeToLog("\n");
-    flushLog();
+    unlikely_if(!fmt) {
+        return;
+    }
+
+    char buf[BSG_KSLOGGER_CBufferSize];
+    const int size = sizeof(buf);
+
+    int len = snprintf(buf, size, "%s: %s:%u: %s(): ",
+                       level, lastPathEntry(file), line, function);
+    if (len < 0) {
+        return;
+    }
+    if (len > size) {
+        len = size;
+    }
+
+    if (len < size - 1) {
+        va_list args;
+        va_start(args, fmt);
+        int max = size - len;
+        int msglen = vsnprintf(buf + len, max, fmt, args);
```

Comment:
If msglen < 0 there was an error and we should just return without writing.

Reply:
Done 👍